src/boot_O2/gfxprint.c

```diff
@@ -93,49 +86,95 @@ void GfxPrint_PrintCharImpl(GfxPrint* this, u8 c) {
+        gDPSetColor(this->dlist++, G_SETPRIMCOLOR, this->color.rgba);
     }
 
-    gSPTextureRectangle(this->dlist++, this->posX, this->posY, this->posX + 32, this->posY + 32, (u16)(tile & 7),
-                        (u16)(c & 4) * 64, (u16)(c >> 3) * 256, 1024, 1024);
+    gSPTextureRectangle(this->dlist++, this->posX, this->posY, this->posX + 32, this->posY + 32, tile, s << 6, t << 8,
+                        1 << 10, 1 << 10);
 
     this->posX += 32;
 }
-#else
-#pragma GLOBAL_ASM("asm/non_matchings/boot/gfxprint/GfxPrint_PrintCharImpl.s")
-#endif
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/gfxprint/GfxPrint_PrintChar.s")
+void GfxPrint_PrintChar(GfxPrint* this, u8 c) {
+    if (c == ' ') {
+        this->posX += 32;
+    } else if (c > ' ' && c < 0x7F) {
+        GfxPrint_PrintCharImpl(this, c);
+    } else if (c >= 0xA0 && c < 0xE0) {
+        if (this->flags & GFXP_FLAG_HIRAGANA) {
+            if (c < 0xC0) {
+                c -= 0x20;
+            } else {
+                c += 0x20;
+            }
+        }
+        GfxPrint_PrintCharImpl(this, c);
+    } else {
+        switch (c) {
+            case '\0':
+                break;
+            case '\n':
+                this->posY += 32;
+            case '\r':
+                this->posX = this->baseX;
+                break;
+            case '\t':
+                do {
+                    GfxPrint_PrintCharImpl(this, ' ');
+                } while ((this->posX - this->baseX) % 256);
+                break;
+            case GFXP_HIRAGANA_CHAR:
+                this->flags |= GFXP_FLAG_HIRAGANA;
+                break;
+            case GFXP_KATAKANA_CHAR:
+                this->flags &= ~GFXP_FLAG_HIRAGANA;
+                break;
+            case GFXP_RAINBOW_ON_CHAR:
+                this->flags |= GFXP_FLAG_RAINBOW;
+                this->flags |= GFXP_FLAG_UPDATE;
+                break;
+            case GFXP_RAINBOW_OFF_CHAR:
+                this->flags &= ~GFXP_FLAG_RAINBOW;
+                this->flags |= GFXP_FLAG_UPDATE;
+                break;
+            case GFXP_UNUSED_CHAR:
+            default:
+                break;
+        }
+    }
+}
 
 void GfxPrint_PrintStringWithSize(GfxPrint* this, const void* buffer, size_t charSize, size_t charCount) {
     const char* str = (const char*)buffer;
     size_t count = charSize * charCount;
 
-    while (count) {
+    while (count != 0) {
         GfxPrint_PrintChar(this, *str++);
         count--;
     }
 }
 
 void GfxPrint_PrintString(GfxPrint* this, const char* str) {
-    while (*str) {
-        GfxPrint_PrintChar(this, *(str++));
+    while (*str != '\0') {
+        GfxPrint_PrintChar(this, *str++);
     }
 }
 
-GfxPrint* GfxPrint_Callback(GfxPrint* this, const char* str, size_t size) {
+void* GfxPrint_Callback(void* arg, const char* str, size_t size) {
+    GfxPrint* this = arg;
+
     GfxPrint_PrintStringWithSize(this, str, sizeof(char), size);
     return this;
 }
```

Comment:
```suggestion
    GfxPrint_PrintStringWithSize(this, str, sizeof(char), size);

    return this;
}
```
Same silly nitpick comparing to OoT, I like the extra space. Not required though